File: frontend/src/api/llmConfig.ts
import { get, put, post } from './http'

export interface LlmConfig {
  provider: string
  base_url: string
  model: string
  active: boolean
  api_key_masked: string
  is_set: boolean
  source: string // default | global
  fallback_reason: string
  timeout_seconds: number
  max_retries: number
  temperature: number
}

export interface LlmTestResult {
  success: boolean
  message: string
  model: string
  duration_ms: number
  attempts: number
  retryable: boolean
  next_action: string
}

export interface LlmDraft {
  provider?: string
  base_url?: string
  model?: string
  api_key?: string
  timeout_seconds?: number
  max_retries?: number
  temperature?: number
}

export interface LlmModelsResult {
  success: boolean
  message: string
  models: string[]
  selected_model: string
  duration_ms: number
  attempts: number
  fallback: boolean
  retryable: boolean
  next_action: string
}

/** 获取全局 LLM 配置 (管理员) */
export function getLlmConfig(): Promise<LlmConfig> {
  return get<LlmConfig>('/admin/llm/config')
}

/** 更新全局 LLM 配置 (管理员) */
export function updateLlmConfig(data: Partial<{
  provider: string
  base_url: string
  model: string
  api_key: string
  active: boolean
  timeout_seconds: number
  max_retries: number
  temperature: number
}>): Promise<LlmConfig> {
  return put<LlmConfig>('/admin/llm/config', data)
}

/** 测试连接 (留空字段用已保存配置) */
export function testLlmConfig(data: LlmDraft): Promise<LlmTestResult> {
  return post<LlmTestResult>('/admin/llm/test', data)
}

/** 从当前表单或已保存配置拉取 OpenAI-compatible 模型列表。 */
export function fetchLlmModels(data: LlmDraft): Promise<LlmModelsResult> {
  return post<LlmModelsResult>('/admin/llm/models', data)
}

/* ── 模型注册表与角色分配 ── */

export interface ModelRegistryItem {
  id: string
  label: string
  vision: boolean
  source: 'pulled' | 'manual'
  capability_source?: 'official' | 'manual' | 'unverified'
  added_at: string
}

export interface ModelRegistryView {
  models: ModelRegistryItem[]
  assignments: Record<string, string>
  roles: Record<string, string>
  warnings?: string[]
}

export interface RegistrySyncResult {
  success: boolean
  message: string
  fetched: string[]
  added: string[]
  models: ModelRegistryItem[]
}

/** 模型注册表 + 角色分配现状 */
export function getModelRegistry(): Promise<ModelRegistryView> {
  return get<ModelRegistryView>('/admin/llm/models/registry')
}

/** 从 provider 拉取最新模型并合并进注册表 */
export function syncModelRegistry(data: LlmDraft): Promise<RegistrySyncResult> {
  return post<RegistrySyncResult>('/admin/llm/models/registry/sync', data)
}

/** 整表保存注册表(手工增删/能力标记) */
export function saveModelRegistry(models: Array<{ id: string; label?: string; vision?: boolean }>): Promise<ModelRegistryView> {
  return put<ModelRegistryView>('/admin/llm/models/registry', { models })
}

/** 更新角色->模型分配(空串清除) */
export function saveModelAssignments(assignments: Record<string, string>): Promise<ModelRegistryView> {
  return put<ModelRegistryView>('/admin/llm/models/assignments', { assignments })
}

File: backend/app/schemas/llm_config.py
"""
全局大模型(LLM)提供商配置 Pydantic Schema(管理员)
"""
from typing import List, Optional

from pydantic import BaseModel, Field


class LlmConfigOut(BaseModel):
    """脱敏输出"""
    provider: str = "deepseek"
    base_url: str = ""
    model: str = ""
    active: bool = False
    api_key_masked: str = ""
    is_set: bool = False
    source: str = "default"  # default=走系统DeepSeek | global=走自定义覆盖
    fallback_reason: str = ""
    timeout_seconds: int = Field(default=60, ge=5, le=600)
    max_retries: int = Field(default=2, ge=0, le=5)
    temperature: float = Field(default=0.2, ge=0, le=2)


class LlmConfigIn(BaseModel):
    """更新全局 LLM 配置;api_key 留空(None)表示不修改,空串表示清空"""
    provider: Optional[str] = Field(default=None, pattern="^(deepseek|openai|custom)$")
    base_url: Optional[str] = Field(default=None, max_length=512)
    model: Optional[str] = Field(default=None, max_length=128)
    api_key: Optional[str] = Field(default=None, max_length=512)
    active: Optional[bool] = None
    timeout_seconds: Optional[int] = Field(default=None, ge=5, le=600)
    max_retries: Optional[int] = Field(default=None, ge=0, le=5)
    temperature: Optional[float] = Field(default=None, ge=0, le=2)


class LlmTestIn(BaseModel):
    """测试连接;字段留空则用已保存配置"""
    provider: Optional[str] = Field(default=None, pattern="^(deepseek|openai|custom)$")
    base_url: Optional[str] = Field(default=None, max_length=512)
    model: Optional[str] = Field(default=None, max_length=128)
    api_key: Optional[str] = Field(default=None, max_length=512)
    timeout_seconds: Optional[int] = Field(default=None, ge=5, le=600)
    max_retries: Optional[int] = Field(default=None, ge=0, le=5)
    temperature: Optional[float] = Field(default=None, ge=0, le=2)


class LlmModelsIn(LlmTestIn):
    """模型发现请求;允许使用尚未保存的表单配置。"""


class LlmModelsOut(BaseModel):
    """模型发现结果;上游不支持列表时可保留手工模型继续。"""
    success: bool
    message: str
    models: List[str] = Field(default_factory=list)
    selected_model: str = ""
    duration_ms: int = 0
    attempts: int = 0
    fallback: bool = False
    retryable: bool = False
    next_action: str = ""
